New Google Drive approach

mbrousil opened the pull request:
Hey B,

What I've done here is to change the Google Drive integration so that files are stored and referenced using folders within Google Drive instead of individual links. I think you suggested this last week, but I forget exactly. Let me know if anything feels underdeveloped or would be difficult to pass between repositories. For example, I'm not sure how best to reference the Google Drive when jumping between repos: Should there be a single link to the overall Drive folder shared between repos, or does it not matter if the users of the second repo have ROSSyndicate Drive access? (Meaning we assume they do and so don't provide any reference link for downloading). "Stable" versions are also a part of this and I updated function documentation.

@matthewross07 if you're interested in trying this workflow out for yourself either now or after this is merged, here's basically how it works:
- A WQP inventory is done using chlorophyll `CharacteristicNames`, an AOI, and a grid intersecting that AOI
- Records per site are counted and used to group the data into appropriately-sized chunks for WQP queries
- Chlorophyll data are downloaded
- The chlorophyll downloads and query arguments are exported to Google Drive. The dataset (not queries) are then downloaded from there to test their portability

mbrousil committed:
Add file uploading and "stable" option

mbrousil committed:
Switch away from stable

mbrousil committed:
Comment

mbrousil committed:
Update with new Google Drive methods

matthewross07 commented:
I think this approach makes a lot of conceptual sense (though I haven't yet run it). Overall this seems like a solid way forward for not only this project, but lots of our larger data/targets projects.

steeleb commented on _targets.R:
```diff
@@ -44,6 +44,26 @@ list(
     )
   ),
   
+  # CharacteristicNames and other query info can change over time, so it can
```
suggestion for reducing manual touches:

Instead of any manual copy/paste into a stable subfolder, could we have a target with an if-statement where there is a setting created in a config file and then add an argument to the `export_single_file()` function? 

it might look like this (the last target is borrowed from your code), these would be the settings we chatted about in Teams (and feel free to rename these):

```
  # settings for pipeline
  # Because targets works in its own session, you need to store the google email
  # you have authenticated with in the run.R script for use in the pipeline
  tar_target(
    name = google_email,
    command = "therossyndicate@gmail.com"
  ),
  
  tar_target(
    name = chl_create_stable_download,
    # we may wish to add some additional flexibility here, like indicating which, 
    # if any parameters this is applicable to - maybe this is only a thing for
    # aquamatch
    command = FALSE # c(TRUE, "chl")
  ),
  
  # Here we choose whether we want a stable version of the `chl_wqp_data`
  # dataset, or to use the dynamic version downloaded above. This choice affects
  # the targets that are read into the pipeline:
  tar_target(
    name = data_version_stable,
    command = TRUE
  ),

```

see suggestions in src/functions.R for implementation!

The only other thing I would consider would be to add some unique identifier for the stable versions, as I don't believe that the link address changes when you upload a new version. If you were to implement a gsub for the download string - this could also be a setting in this section, if we wanted to go that route.

steeleb commented on run.R:
```diff
@@ -31,12 +31,25 @@ lapply(required_pkgs, package_installer)
+
+# Google Drive auth -------------------------------------------------------
+
+# Confirm Google Drive is authorized locally
```
```suggestion
# Confirm Google Drive is authorized locally. If you are a member of the AquaSat team and 
# are updating the stable version of the pipeline, you will need to authorize Google using
# the ROSSyndicate gmail account. All other users can use any google address, allowing 
# for their personal 'version' of the data download for use throughout the remainder of the
# pipeline.
```

steeleb replied:
upon implementation of the google_drive target, we would need to add that the email authorized here needs to match that in the `google_email` target.

Of note, someone without access to that email would not be able to run just the final target, pulling in the stable file, but I think that has an easy solution. (More on that in the _targets.R file)

steeleb commented on README.md:
```diff
@@ -1,5 +1,5 @@
 # targets_storage
 
-This repository is meant to be a small vignette of the `AquaSat/AquaMatch_download_WQP` which documents some pain points in our use of {targets} with large data objects, especially regarding WQP. download, which is not static.
+This repository is meant to be a small vignette of the `AquaSat/AquaMatch_download_WQP` which documents some pain points in our use of {targets} with large data objects, especially regarding WQP download, which is not static.
```
```suggestion
This repository is meant to be a small vignette of the `AquaSat/AquaMatch_download_WQP` which 
documents some pain points in our use of {targets} with large data objects, especially regarding WQP 
download, which is not static. This architecture requires a google account to function, and one intends to 
update the pipeline for the external world (aka, create stable datasets for downstream use, e.g. AquaSat), 
you will need to authenticate using the ROSSyndicate google account.
```

steeleb commented on src/functions.R:
```diff
@@ -227,3 +267,107 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+#' 
+export_single_file <- function(target, drive_path){
+  
+  require(googledrive)
```
```suggestion
  # authorize using the google email provided
  drive_auth(google_email)
```

steeleb commented on src/functions.R:
```diff
@@ -227,3 +267,107 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+#' 
+retrieve_data <- function(target, local_folder, drive_folder, stable, file_type = ".rds"){
+  
+  require(googledrive)
```
```suggestion
  # authorize using the google email provided
  drive_auth(google_email)
```

steeleb commented on src/functions.R:
```diff
@@ -227,3 +267,107 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+#' Returns a local path to a csv file containing a text link to the uploaded
+#' file in Google Drive.
+#' 
+export_single_file <- function(target, drive_path){
```
```suggestion
export_single_file <- function(target, drive_path, stable, google_email){
```

steeleb replied:
you could also set the default as 'stable = FALSE' here if you only want to create stable versions of *some* chl .rds files by only using the argument `stable = chl_create_stable_file` at targets that you actually want to save stable files for.

steeleb commented on src/functions.R:
```diff
@@ -227,3 +267,107 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+#' A function to export a single target (as a file) to Google Drive and return
+#' the shareable Drive link as a file path.
+#' 
+#' @param target The name of the target to be exported (as an object not a string).
```
question: why as the object? This then requires that the target was run, which defeats the purpose of time savings, right? I think this is equally efficient if set as a character string (or, per code within this function 'target_string').

mbrousil replied:
Since this is an export step wouldn't it require running the target no matter what? I might be skipping a step in my head though. And I use object here to define the target inter-dependencies as simply as possible.

steeleb replied:
This is my b. I meant to leave this on the `retrieve_data()` function!! There, it's for the user who just wants to use the stable version and not mess with downloading -- or in our case, where we just pull the .rds into our other pipeline.

steeleb commented on src/functions.R:
```diff
@@ -227,3 +267,107 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+#' 
+#' @param target The name of the target to be exported (as an object not a string).
+#' 
+#' @param folder_pattern A local path to a location where the target should be
```
I think this is an artifact of a previous version, can delete! :)

steeleb commented on src/functions.R:
```diff
@@ -227,3 +267,107 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+                        name = paste0(target_string, ".rds"))
+  
+  # Make the Google Drive link shareable: anyone can view
+  out_file_share <- out_file %>%
```
suggestion: these two lines could be condensed as the following:

```suggestion
drive_share(out_file, role = "reader", type = "anyone")
```

steeleb commented on src/functions.R:
```diff
@@ -227,3 +267,107 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+  # Local file download location
+  local_path <- file.path(local_folder, target_string)
+  
+  # Get file contents of the Google Drive folder specified. If stable == TRUE
```
I think something like this could be added to the `export_single_file()` function, too. I started to code it up and then broke everything on my end. LOL.

steeleb commented on src/functions.R:
```diff
@@ -227,3 +267,107 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+  
+}
+
+#' @title Retrieve a dataset from Google Drive
```
As written, this requires that the user be someone who has access to the ROSS email. For AquaSat, and probably for other projects, we will want to have a link for folks to use. I've documented an example of that in the _targets.R file.

steeleb commented on _targets.R:
```diff
@@ -131,21 +151,34 @@ list(
+    name = chl_wqp_data_link_file,
+    command = export_single_file(target = chl_wqp_data,
+                                 drive_path = "~/targets_storage_example/chlorophyll/")
+  ),
```
Re: my note in run.R --

I think we need one more target here (or maybe a pair of targets), that uses the drive_ls() funciton to store the google drive `as_id()` for this file, that is also tracked by GitHub (this would be the equivalent of your saving the copy/paste https in the previous iteration). This would enable downstream users who we don't want to have the ROSS email authentication (aka, the general public), to be able to run `tar_make(chl_wqp_data_in)` and have the data ready-to-go with a specific target.

steeleb commented on run.R:
```diff
@@ -31,12 +31,25 @@ lapply(required_pkgs, package_installer)
+2
 
 
 # Run pipeline ------------------------------------------------------------
```
Suggestion: add some details about when to run `tar_make()` versus `tar_make(chl_wqp_data_in)` maybe something along the lines of:

```suggestion
# There are a few options for running this pipeline, and you will need to assure that your Google email has been added in the `config.yml` file. 

# create a fresh version of all {targets} - make sure that line 22 in `_targets.R` is set to "create_new_version"
tar_make()

#  just load the stable version from AquaSat - make sure that line 22 in `_targets.R` is set to "use_stable_link"
tar_make(chl_wqp_data_in)

# members of the AquaSat team can update the stable version using "admin_update" on line 22 in `_targets.R`
tar_make()
```

steeleb commented on src/functions.R:
```diff
@@ -227,3 +267,107 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+#' @returns 
+#' The dataset after being downloaded and read into the pipeline from Google Drive.
+#' 
+retrieve_data <- function(target, local_folder, drive_folder, stable, file_type = ".rds"){
```
```suggestion
retrieve_data <- function(target, local_folder, drive_folder, stable, google_email, file_type = ".rds"){
```

steeleb commented on _targets.R:
```diff
@@ -131,21 +151,34 @@ list(
+  tar_target(
+    name = chl_wqp_data_link_file,
+    command = export_single_file(target = chl_wqp_data,
+                                 drive_path = "~/targets_storage_example/chlorophyll/")
```
```suggestion
                                 drive_path = "~/targets_storage_example/chlorophyll/"),
                                 google_email = google_email,
                                 stable = chl_create_stable_download),
    packages = "googledrive"
```

steeleb commented on _targets.R:
```diff
@@ -44,6 +44,26 @@ list(
+  tar_target(
+    name = chl_wqp_characteristics_file,
+    command = export_single_file(target = chl_wqp_characteristics,
+                                 drive_path = "~/targets_storage_example/chlorophyll/")
```
```suggestion
                                 drive_path = "~/targets_storage_example/chlorophyll/",
                                 stable = chl_create_stable_download,
                                 google_email = google_email),
  packages = "googledrive"
```

steeleb commented on _targets.R:
```diff
@@ -44,6 +44,26 @@ list(
+  tar_target(
+    name = wqp_args_file,
+    command = export_single_file(target = wqp_args,
+                                 drive_path = "~/targets_storage_example/chlorophyll/")
```
```suggestion
                                 drive_path = "~/targets_storage_example/chlorophyll/",
                                 stable = chl_create_stable_download,
                                 google_email = google_email),
  packages = "googledrive"
```

steeleb reviewed (commented):
SUHWEET.

Lots of suggestions here, and I think I provided code snips for all of them. If anything is confusing or I'm absolutely insane, please let me know, happy to elucidate as needed. 

To merge, please address the addition of the settings - at least of the google email. Note, in my suggested edits, I moved the `data_version_stable` target to the top for end-user ease. 

Two additional items that I'd really like implemented are:

- the add of the targets to get/save the as_id() for the chl data file so that one can run `tar_make(chl_data_file_in)` in a more standalone fashion and such that you could run it without being an authorized user of the ROSS email.
- and stable files saved in Drive within the stable folder without manual intervention (I made suggestions about how that might be accomplished)

If these two are better for individual PRs outside of this, that's totally cool by me.

mbrousil committed:
Apply suggestions from code review

Co-authored-by: B Steele <32140074+steeleb@users.noreply.github.com>

mbrousil committed:
Add ideas from PR review

mbrousil commented:
@steeleb I went through and implemented what you suggested, but on occasion in slightly different ways. An overview of the changes:
- Added a config yaml using {config} for email and stable options
- Exporting files requires email and stability inputs
- Exported files to Drive include date stamps in their name
- Added a target (and function) to grab file ids from Drive for a specified folder

steeleb commented on _targets.R:
```diff
@@ -15,9 +15,18 @@ tar_source("src/")
 list(
   
+  # Grab configuration information for the workflow run (config.yml)
+  tar_target(
```
this is niiiiiiiice. time to go implement in my other repos.

steeleb commented on src/functions.R:
```diff
@@ -227,3 +268,231 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+                        name = paste0(target_string, ".rds"))
+  
+  # Make the Google Drive link shareable: anyone can view
+  drive_share(out_file, role = "reader", type = "anyone")
```
```suggestion
  drive_share_anyone(out_file)
```

There's a wrapper for this that I accidentally found the other day! :)

steeleb commented on src/functions.R:
```diff
@@ -227,3 +268,231 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+  drive_share(out_file, role = "reader", type = "anyone")
+  
+  # If stable == TRUE then export a second, dated file to the stable/ subfolder
+  if(stable){
```
heck yes.

steeleb commented on src/functions.R:
```diff
@@ -227,3 +268,231 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+#' A function to retrieve a dataset from Google Drive after it has been uploaded
+#' in a previous step.
+#' 
+#' @param target The name of the target to be exported (as an object not a string).
```
```suggestion
#' @param target The name of the target to be retreived (as an object not a string).
```

steeleb replied:
This is the one that I think should be a string, not an object that [I misplaced here](https://github.com/rossyndicate/targets_storage/pull/3/files#r1496557980). This is primarily so that it doesn't require that the whole workflow is run if `chl_stable: TRUE`.

Basically if the config is set `chl_stable = TRUE` and `chl_create_stable = FALSE`, we could run just `tar_make(chl_wqp_data_in)` and it's super quick.

steeleb commented on src/functions.R:
```diff
@@ -227,3 +268,231 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+#' @param target The name of the target to be exported (as an object not a string).
+#' 
+#' @param local_folder A string specifying the folder where the file should be
+#' downloaded.
```
```suggestion
#' downloaded to.
```

steeleb commented on src/functions.R:
```diff
@@ -227,3 +268,231 @@ fetch_wqp_data <- function(site_counts_grouped, char_names, wqp_args = NULL,
+}
+
+
+#' @title Function to retrieve Google Drive IDs for a specified folder.
```
Comment/suggestion: My initial read of this function is that it's overly complicated in a way that it doesn't need to be. Here's some context as to why: this function will be used to populate a .csv that's tracked even though it's not currently (`data/out/chl_stable_drive_ids.csv`) so that downstream users (or us in a different repo) can easily download from Drive without running the entire workflow. If you set a specific version date (either as a config setting, or as a general argument in the function), this function becomes much more straightforward and you can just reduce the dribble by the pattern of the version setting.

steeleb commented on config.yml:
```diff
@@ -0,0 +1,6 @@
+default:
```
Suggest adding a couple of simple options here... Note, I haven't tried these to see if they throw errors
```suggestion
use_stable_links:
   google_email: ""
   chl_stable: TRUE
   chl_create_stable: FALSE
create_new_version:
   google_email: ""
   chl_stable: FALSE
   chl_create_stable: FALSE
admin_update:
```

steeleb commented on _targets.R:
```diff
@@ -15,9 +15,18 @@ tar_source("src/")
+  tar_target(
+    name = workflow_config,
+    # The config package does not like to be used with library()
+    command = config::get()
```
```suggestion
    command = config::get(config = "admin_update")
```

steeleb reviewed (commented):
rockstar. this is awesome. I have just a small tweak to suggest:

1) track that one out file in the .gitignore for downstream users:
!!data/out/chl_stable_drive_ids.csv

And then a bigger tweak(?), that I'll let you ruminate on:
2) Consider using a 'version_date' or 'version_id' setting in the config file so that you aren't reliant on Sys.Date() - this is actually an issue I've faced with processing that occurs over multiple days (like the AquaSat/lakeSR workflow) and then I miss some data when I pull info (or get too much data because I don't filter by date and have multiple near-duplicates). If you use that specified version date (that would only be for updating the stable files), you can also use that as a setting for download version in retrieve_data(). 

I also added some suggested config settings and direction to users for the use cases I forsee with this type of a setup.

mbrousil committed:
Apply suggestions from code review

Co-authored-by: B Steele <32140074+steeleb@users.noreply.github.com>

mbrousil committed:
simplify file id function

mbrousil committed:
Add stable drive ids file

mbrousil commented:
> And then a bigger tweak(?), that I'll let you ruminate on: 2) Consider using a 'version_date' or 'version_id' setting in the config file so that you aren't reliant on Sys.Date() - this is actually an issue I've faced with processing that occurs over multiple days (like the AquaSat/lakeSR workflow) and then I miss some data when I pull info (or get too much data because I don't filter by date and have multiple near-duplicates). If you use that specified version date (that would only be for updating the stable files), you can also use that as a setting for download version in retrieve_data().
> 
> I also added some suggested config settings and direction to users for the use cases I forsee with this type of a setup.

@steeleb I thought about this too...I wasn't super concerned about it in the end though since these things will likely be run infrequently and this is mostly an example. Do you think it would be likely to be impactful in practice?

steeleb commented:
I think for this use case, no, this totally works and it's not required. For the greater AquaSat repo (and for my end on lakeSR) it might be useful due to being able to define a version when it processes over multiple days which would result in multiple version dates across the targets that you save to .rds that are from the same run. If you don't add that here, I do think it would be worthwhile to document the case in which you might want to create a config.yml setting for the version date. 

Whichever way you choose, I feel like this is g2g and merge in with main.
